pallets/msa/src/lib.rs

```diff
@@ -399,6 +409,44 @@ pub mod pallet {
 		fn on_initialize(current: T::BlockNumber) -> Weight {
 			Self::reset_virtual_bucket_if_needed(current)
 		}
+
+		fn offchain_worker(_block_number: T::BlockNumber) {
+			let filtered_events: Vec<Event<T>> =
+				<frame_system::Pallet<T>>::read_events_no_consensus()
```

Comment:
I hope we find a better way than to scrape the whole runtime via <frame_system::Pallet<T>>::read_events_no_consensus()

and lot of uncertainties/questions around its usage

Reply:
Same thoughts! this pretty much scrape the wasm state whenever it is called, I am not sure how deterministic offchain_workers are that they dont miss out on states

Not sure how we can get events via block_number internally if anyone has a suggestion

Reply:
One way to do is is using `offchain_index::set` in extrinsic and read on offchain worker, 

however the only way to exchange data directly from runtime and offchain worker using index will require unique keys for each event per block_number to be set in index

which then somehow we scraped at offchain_worker side and we dont want to do mutate onchain for sure